modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1836,11 +1835,12 @@ private CompletableFuture<Void> handleChangePendingAssignmentEvent(
                     )
                     .thenComposeAsync(unused -> inBusyLock(busyLock, () -> {
                         if (!isRecovery) {
+                            HybridTimestamp lwm = lowWatermark.getLowWatermark();
                             // We create index storages (and also register the necessary structures) for the rebalancing one partition
                             // before start the raft node, so that the updates that come when applying the replication log can safely
                             // update the indexes. On recovery node, we do not need to call this code, since during restoration we start
                             // all partitions and already register indexes there.
-                            registerIndexesToTable(tbl, catalogService, singlePartitionIdSet, tbl.schemaView());
+                            registerIndexesToTable(tbl, catalogService, singlePartitionIdSet, tbl.schemaView(), lwm);
```

Comment:
Here it looks like there may be a race, for example, a low watermark will be aggressively updated quite quickly and it may turn out that the task of destroying the index will end before we register the index and it will be created empty.

Reply:
Yep. That's why I like the way destroying tables/indexes via Catalog event.

Reply:
A ticket created.
https://issues.apache.org/jira/browse/IGNITE-21738